File: app/src/main/java/id/ugm/ahpsaw/data/MatriksElemen.java
package id.ugm.ahpsaw.data;

import java.io.Serializable;

public class MatriksElemen implements Serializable {
    double[][] comparison;
    double[] weight;
    int n;
    double weightOfParent;
    String[] nameOfCriteria;


    public MatriksElemen(double[][] comparison, String[] name) {
        this.comparison = comparison;
        this.nameOfCriteria = name;
        this.n = name.length;
        this.weight = new double[n];
        this.weightOfParent = 1;
    }
    public MatriksElemen(double[][] comparison, String[] name, double weightOfParent) {
        this.comparison = comparison;
        this.nameOfCriteria = name;
        this.n = name.length;
        this.weight = new double[n];
        this.weightOfParent = weightOfParent;
    }

    private double[] calcWeight() {
        double[][] norm = new double[n][n];
        double[] sumColumn = new double[n];
        double[] sumRow = new double[n];
        //NORMALIZATION

        for (int i = 0; i < n; i++) {
            sumColumn[i] = 0.0;
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                sumColumn[j] += comparison[i][j];
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                norm[i][j] = comparison[i][j] / sumColumn[j];
                //System.out.println("norm[" + i + "][" + j +"]= "+ norm[i][j]);
            }
        }

        //CALCULATE WEIGHT
        for (int i = 0; i < n; i++) {
            sumRow[i] = 0.0;
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                sumColumn[j] += comparison[i][j];
            }
        }
        for (int i = 0; i < n; i++) {
            sumRow[i] = 0.0;
            for (int j = 0; j < n; j++) {
                sumRow[i] += norm[i][j];
            }
            //System.out.println("sumRow" + i +"= " + sumRow[i]);
            weight[i] = sumRow[i] / n;
        }
        return weight;
    }

    public double[] calcAbsoluteWeight() {
        double[] localWeight = calcWeight();
        double[] absoluteWeight = new double[calcWeight().length];
        for (int i=0; i<calcWeight().length; i++){
            absoluteWeight[i] = localWeight[i]*weightOfParent;
        }
        return absoluteWeight;
    }

    public double consistencyRatio() {
        double[] WSV = new double[n];
        double[] CV = new double[n];
        double eigenMaks;
        double CI;
        double CR;
        double[] RC = new double[]{0.0, 0.0, 0.58, 0.9, 1.12, 1.24, 1.32, 1.41, 1.45, 1.49, 1.51, 1.48, 1.56, 1.57, 1.59};

        //CALCULATE WSV
        calcWeight();
        for (int i = 0; i < n; i++) {
            WSV[i] = 0.0;
            for (int j = 0; j < n; j++) {
                WSV[i] += comparison[i][j] * weight[j];
            }
        }

        //CALCULATE CV
        for (int i = 0; i < n; i++) {
            CV[i] = WSV[i] / weight[i];
        }

        //CALCULATE EIGENMAKS
        double sumCV = 0.0;
        for (int i = 0; i < n; i++) {
            sumCV += CV[i];
        }
        eigenMaks = sumCV / n;

        //CALCULATE CI
        CI = (eigenMaks - n) / (n - 1);

        //CALCULATE CR
        CR = CI / RC[n - 1];
        return CR;
    }

    public String[] getNameOfCriteria() {
        return nameOfCriteria;
    }

    public void setNameOfCriteria(String[] nameOfCriteria) {
        this.nameOfCriteria = nameOfCriteria;
    }

    public double[][] getComparison() {
        return comparison;
    }

    public void setComparison(double[][] comparison) {
        this.comparison = comparison;
    }
}

File: app/src/main/java/id/ugm/ahpsaw/data/MatriksElemen.kt
package id.ugm.ahpsaw.data

import android.util.Log
import java.io.Serializable

class MatriksElemen : Serializable {
    var comparison: Array<DoubleArray>
    var weight: DoubleArray
    var n: Int
    var weightOfParent: Double
    var nameOfCriteria: Array<String>

    constructor(comparison: Array<DoubleArray>, name: Array<String>) {
        this.comparison = comparison
        nameOfCriteria = name
        n = name.size
        weight = DoubleArray(n)
        weightOfParent = 1.0
        comparison.forEachIndexed{i, item ->
            item.forEachIndexed{j, item2 ->
                Log.i("MatriksElemen Construction(no parent weight): ", "matriks[$i][$j]=${comparison[i][j]}")
            }
        }

    }

    constructor(
        comparison: Array<DoubleArray>,
        name: Array<String>,
        weightOfParent: Double
    ) {
        this.comparison = comparison
        nameOfCriteria = name
        n = name.size
        weight = DoubleArray(n)
        this.weightOfParent = weightOfParent
        comparison.forEachIndexed{i, item ->
            item.forEachIndexed{j, item2 ->
                Log.i("MatriksElemen Construction(WITH PARENT WEIGHT): ", "matriks[$i][$j]=${comparison[i][j]}")
            }
        }
    }

    private fun calcWeight(): DoubleArray {
        val norm = Array(n) { DoubleArray(n) }
        val sumColumn = DoubleArray(n)
        val sumRow = DoubleArray(n)
        //NORMALIZATION
        for (i in 0 until n) {
            sumColumn[i] = 0.0
        }
        for (i in 0 until n) {
            for (j in 0 until n) {
                sumColumn[j] += comparison[i][j]
            }
        }
        for (i in 0 until n) {
            for (j in 0 until n) {
                norm[i][j] = comparison[i][j] / sumColumn[j]
                //System.out.println("norm[" + i + "][" + j +"]= "+ norm[i][j]);
            }
        }

        //CALCULATE WEIGHT
        for (i in 0 until n) {
            sumRow[i] = 0.0
        }
        for (i in 0 until n) {
            for (j in 0 until n) {
                sumColumn[j] += comparison[i][j]
            }
        }
        for (i in 0 until n) {
            sumRow[i] = 0.0
            for (j in 0 until n) {
                sumRow[i] += norm[i][j]
            }
            //System.out.println("sumRow" + i +"= " + sumRow[i]);
            weight[i] = sumRow[i] / n
        }
        return weight
    }

    fun calcAbsoluteWeight(): DoubleArray {
        val localWeight = calcWeight()
        val absoluteWeight = DoubleArray(calcWeight().size)
        for (i in 0 until n) {
            absoluteWeight[i] = localWeight[i] * weightOfParent
            Log.i("calcAbsoluteWeight(): ", "absoluteWeight[$i]=${localWeight[i]} * ${weightOfParent} = ${absoluteWeight[i]}")
        }
        return absoluteWeight
    }

    fun consistencyRatio(): Double {
        val WSV = DoubleArray(n)
        val CV = DoubleArray(n)
        val eigenMaks: Double
        val CI: Double
        val CR: Double
        val RC = doubleArrayOf(
            0.0,
            0.0,
            0.58,
            0.9,
            1.12,
            1.24,
            1.32,
            1.41,
            1.45,
            1.49,
            1.51,
            1.48,
            1.56,
            1.57,
            1.59
        )

        //CALCULATE WSV
        calcWeight()
        for (i in 0 until n) {
            WSV[i] = 0.0
            for (j in 0 until n) {
                WSV[i] += comparison[i][j] * weight[j]
            }
        }

        //CALCULATE CV
        for (i in 0 until n) {
            CV[i] = WSV[i] / weight[i]
        }

        //CALCULATE EIGENMAKS
        eigenMaks = CV.sum()/ n

        //CALCULATE CI
        CI = (eigenMaks - n) / (n - 1)

        //CALCULATE CR
        CR = CI / RC[n - 1]
        return CR
    }

}